### PR TITLE
ProgLED and ProgButton pins are now optional

### PR DESCRIPTION
**Pull request by Adminius**
New init function for own ProgLED and ProgButton implementation
New DemoSketch for new init-function
DemoSketches updated to beta4a

### CODE REVIEW

File: src/KonnektingDevice.cpp [hunk 1]
```diff
@@ -123,16 +123,111 @@ KonnektingDevice::KonnektingDevice() {
...
+ *          The ID of the device
+ *  @param  revisionID
+ *          The device's revision
+ *  @return void
+ */
+
+/**************************************************************************/
+
+void KonnektingDevice::init(HardwareSerial& serial,
+        void (*func)(bool),
```
Comment (tuxedo0801) [inline src/KonnektingDevice.cpp]: the variable's name is "func". better name would be "progIndicatorFunc". The variable name must/should match the name in the doxygen doc. See above.

File: src/KonnektingDevice.cpp [hunk 2]
```diff
@@ -123,16 +123,111 @@ KonnektingDevice::KonnektingDevice() {
 
 }
 
-
 /**************************************************************************/
 /*!
  *  @brief   Starts KNX KonnektingDevice, as well as KNX Device
  *  @param  serial 
  *          serial port reference, f.i. "Serial" or "Serial1"
+ *  @param  setProgLedFunc
```
Comment (tuxedo0801) [inline src/KonnektingDevice.cpp]: Param name here must match param name in method-signature. See below!

File: src/KonnektingDevice.cpp [hunk 3]
```diff
@@ -123,16 +123,111 @@ KonnektingDevice::KonnektingDevice() {
...
+
+    DEBUG_PRINTLN("15/7/255 = 0x%04x", G_ADDR(15, 7, 255));
+
+    _initialized = true;
+
+    _manufacturerID = manufacturerID;
+    _deviceID = deviceID;
+    _revisionID = revisionID;
+
+    setProgLedFunc = func;
```
Comment (tuxedo0801) [inline src/KonnektingDevice.cpp]: global variables are prefixed with "_". Name it f.i. "_progIndicatorFunc".

File: src/KonnektingDevice.cpp [hunk 4]
```diff
@@ -123,16 +123,111 @@ KonnektingDevice::KonnektingDevice() {
...
+ *  @param  deviceID
+ *          The ID of the device
+ *  @param  revisionID
+ *          The device's revision
+ *  @return void
+ */
+
+/**************************************************************************/
+
+void KonnektingDevice::init(HardwareSerial& serial,
```
Comment (tuxedo0801) [inline src/KonnektingDevice.cpp]: Looks like you just duplicated the init() function with all it's code and added the new argument. 

Try to have kind of "internal init" function that is called from both init() variants. That saves memory and makes code more reusable.

File: src/KonnektingDevice.cpp [hunk 5]
```diff
@@ -169,16 +264,10 @@ void KonnektingDevice::init(HardwareSerial& serial,
 
     pinMode(_progLED, OUTPUT);
     pinMode(_progButton, INPUT);
-    //digitalWrite(_progButton, HIGH); //PULLUP
 
-    digitalWrite(_progLED, LOW);
+    setProgState(false);
 
-#ifdef ARDUINO_ARCH_SAMD
```
Comment (tuxedo0801) [inline src/KonnektingDevice.cpp]: did you verify this? Last time I tried this, it did  not work without this special treatment.

File: src/KonnektingDevice.cpp [hunk 6]
```diff
@@ -231,6 +320,9 @@ void KonnektingDevice::init(HardwareSerial& serial,
         delay(500);
         reboot();
     }
+#ifdef ESP32
+    EEPROM.begin(1024);
```
Comment (tuxedo0801) [inline src/KonnektingDevice.cpp]: ESP8266 does this in KonnektingDevice Constructor. Wouldn't it make sense to do the same for ESP32?

File: src/KonnektingDevice.h [hunk 7]
```diff
@@ -161,14 +161,16 @@ class KonnektingDevice {
...
+
     int getFreeEepromOffset();
 
 private:
 
     bool _rebootRequired = false;
     bool _initialized = false;
 
-    int _progbtnCount = 0;
+    byte _progbtnCount = 0;
```
Comment (tuxedo0801) [inline src/KonnektingDevice.h]: maybe add ifdef for REBOOT_BUTTON ?? That way it's not in  memory if it's not used at all.

**Review by tuxedo0801 — CHANGES_REQUESTED** [review]
check my comments.

File: src/KonnektingDevice.cpp (hunk 1 above)
Comment (Adminius) [reply src/KonnektingDevice.cpp]: func is temporary variable.
setProgLedFunc renamed to _progIndicatorFunc

File: src/KonnektingDevice.cpp (hunk 4 above)
Comment (Adminius) [reply src/KonnektingDevice.cpp]: new internal init function added

File: src/KonnektingDevice.cpp (hunk 5 above)
Comment (Adminius) [reply src/KonnektingDevice.cpp]: yes, it works

File: src/KonnektingDevice.cpp (hunk 6 above)
Comment (Adminius) [reply src/KonnektingDevice.cpp]: fixed

File: src/KonnektingDevice.h (hunk 7 above)
Comment (Adminius) [reply src/KonnektingDevice.h]: #ifdef added

File: src/KonnektingDevice.cpp (hunk 3 above)
Comment (Adminius) [reply src/KonnektingDevice.cpp]: fixed, also for old eeprom functions

File: src/KonnektingDevice.cpp (hunk 2 above)
Comment (Adminius) [reply src/KonnektingDevice.cpp]: fixed

**Review by tuxedo0801 — APPROVED** [review]
looks good

**Comment by tuxedo0801** [comment]
There are conflicts due to my changes in the beta4 branch to get travis CI up&running.
The most conflicts come from the examples, and there is one conflict in knxDevice.cpp (I guess due to bool vs. boolean change).

Can you pls. pull these changes and merge them into your branch to solve the conflicts?

**Pull request merged**